tiflash/tune-tiflash-performance.md

```diff
@@ -1,59 +1,103 @@
+
+If the aggregate function with the `distinct` operation is slow in the query, you can enable the optimizer to execute the optimization operation of pushing down the aggregate function with `Distinct` (such as `select sum(distinct a) from t`) to Coprocessor by setting the value of the [`tidb_opt_distinct_agg_push_down`](/system-variables.md#tidb_opt_distinct_agg_push_down) variable to `ON`.
+
+```sql
+set @@tidb_opt_distinct_agg_push_down = ON;
+```
+
+### Compact data using the `ALTER TABLE ... COMPACT` statement
+
+Executing the [`ALTER TABLE ... COMPACT`](/sql-statements/sql-statement-alter-table-compact.md) statement can initiate compaction for a specific table or partition on a TiFlash node. During the compaction, storage nodes rewrite physical data, including cleaning up deleted rows and merging multiple versions of data caused by updates. This helps enhance read performance and reduce disk usage. The following are examples:
 
-5. Compact data using the `ALTER TABLE ... COMPACT` statement if necessary:
+```sql
+ALTER TABLE employees COMPACT TIFLASH REPLICA;
+```
 
-    Executing the [`ALTER TABLE ... COMPACT`](/sql-statements/sql-statement-alter-table-compact.md) statement can initiate compaction for a specific table or partition on a TiFlash node. During the compaction, storage nodes rewrite physical data, including cleaning up deleted rows and merging multiple versions of data caused by updates. This helps enhance read performance and reduce disk usage. The following are examples:
+```sql
+ALTER TABLE employees COMPACT PARTITION pNorth, pEast TIFLASH REPLICA;
+```
+
+### Replace Shuffled Hash Join with Broadcast Hash Join
+
+For `Join` operations with small tables, the Broadcast Hash Join algorithm can avoid transmitting large tables, thereby improving the compute performance.
+
+- The [`tidb_broadcast_join_threshold_size`](/system-variables.md#tidb_broadcast_join_threshold_size-new-in-v50) variable controls whether to use the Broadcast Hash Join algorithm. If the table size (unit: byte) is smaller than the value of this variable, the Broadcast Hash Join algorithm is used. Otherwise, the Shuffled Hash Join algorithm is used.
 
     ```sql
-    ALTER TABLE employees COMPACT TIFLASH REPLICA;
+    set @@tidb_broadcast_join_threshold_size = 2000000;
     ```
 
+- The [`tidb_broadcast_join_threshold_count`](/system-variables.md#tidb_broadcast_join_threshold_count-new-in-v50) variable also controls whether to use the Broadcast Hash Join algorithm. If the objects of the join operation belong to a subquery, the optimizer cannot estimate the size of the subquery result set. In this situation, the size is determined by the number of rows in the result set. If the estimated number of rows in the subquery is fewer than the value of this variable, the Broadcast Hash Join algorithm is used. Otherwise, the Shuffled Hash Join algorithm is used.
```

Comment:
```suggestion
- The [`tidb_broadcast_join_threshold_count`](/system-variables.md#tidb_broadcast_join_threshold_count-new-in-v50) variable also controls whether to use the Broadcast Hash Join algorithm. If the objects of the join operation belong to a subquery, the optimizer cannot estimate the size of the subquery result set. In this situation, the size is determined by the number of rows in the result set. If the estimated number of rows for the subquery is fewer than the value of this variable, the Broadcast Hash Join algorithm is used. Otherwise, the Shuffled Hash Join algorithm is used.
```